Add option to list config files with pip config

deveshks opened the pull request:
Fixes and closes #6741 

Add option to list config files per variant and whether they actually exist

```
$ pip config list-files
env:
global:
  /Library/Application Support/pip/pip.conf, exists: False 
site:
  /Users/devesh/pip/.env/pip.conf, exists: True 
user:
  /Users/devesh/.pip/pip.conf, exists: False 
  /Users/devesh/.config/pip/pip.conf, exists: False 
```

This is a preliminary output, but we need to decide on some things before we finalize something

Edit: An updated format with all values for a config file listed under it is now added:

If `PIP_CONFIG_FILE` is set
```
$ pip config debug
env_var:
  PIP_CONFIG_FILE='/Users/devesh/Desktop/pip.cfg'
  PIP_DEFAULT_TIMEOUT='60'
env:
  /Users/devesh/Desktop/pip.cfg, exists: True
    global.timeout: 60
    global.index-url: https://download.zope.org/ppix
global:
  /Library/Application Support/pip/pip.conf, exists: True
    global.timeout: 60
    freeze.timeout: 10
site:
  /Users/devesh/pip/.env/pip.conf, exists: True
    global.no-cache-dir: false
    install.no-compile: no
    install.no-warn-script-location: false
```

If `PIP_CONFIG_FILE` is not set:

```
$ pip config debug
env_var:
  PIP_DEFAULT_TIMEOUT='60'
env:
global:
  /Library/Application Support/pip/pip.conf, exists: True
    global.timeout: 60
    freeze.timeout: 10
site:
  /Users/devesh/pip/.env/pip.conf, exists: True
    global.no-cache-dir: false
    install.no-compile: no
    install.no-warn-script-location: false
    a.b: 1
user:
  /Users/devesh/.pip/pip.conf, exists: False
  /Users/devesh/.config/pip/pip.conf, exists: True
    global.find-links: 
    http://download.example.com
    install.find-links: 
    http://mirror1.example.com
    http://mirror2.example.com
    install.trusted-host: 
    http://mirror1.example.com
    http://mirror2.example.com
```

gutsytechster commented on src/pip/_internal/commands/configuration.py:
```diff
@@ -180,6 +184,16 @@ def unset_name(self, options, args):
 
         self._save_configuration()
 
+    def list_config_files(self, options, args):
```
You can add the type annotation for this method. :)

deveshks replied:
Actually no other functions have been type-annotated here. I have 3 other PR's adding type annotations to other files in `commands`, probably will create one for this file in near future.

gutsytechster commented on src/pip/_internal/configuration.py:
```diff
@@ -370,7 +370,7 @@ def _get_environ_vars(self):
 
     # XXX: This is patched in the tests.
-    def _iter_config_files(self):
+    def iter_config_files(self):
```
I am not sure if this would be helpful, but you may as well create  `iter_config_files()` which calls the `_iter_config_files()`. This way you won't need to modify the private method definition, rather you'd be exposing its functionality via some other public method.

It's just a suggestion. I'll look forward to other opinions on this.

deveshks replied:
That's a good idea. I actually used `iter_config_files()` as per the suggestion from @pradyunsg from the issue https://github.com/pypa/pip/issues/6741#issuecomment-616101245 . 

 IMO we are using this function outside the `pip._internal.Configuration` module just like we are using `load` (we also plan to expose this in `pip debug` output), so it might be better to rename the function then write another wrapper function and increase the maintenance burden. But yes I will too wait for other opinions on this as well

pradyunsg replied:
There's not much value to add a method that just calls another method, given that we control both the consumers of this API and this API.

deveshks replied:
Agreed. I will just leave the conversion of `_iter_config_files` to `iter_config_files` in this PR as is.

deveshks commented on src/pip/_internal/commands/configuration.py:
```diff
@@ -93,7 +96,8 @@ def run(self, options, args):
             "set": self.set_name_value,
-            "unset": self.unset_name
+            "unset": self.unset_name,
+            "list-files": self.list_config_files
```
Fixed in latest commit.

deveshks commented:
To move the PR ahead, I have implemented the first suggestion provided in https://github.com/pypa/pip/issues/6741#issuecomment-617602424. (Naming the command as `pip config debug` and following it's output.

The output now looks like:

```
$ pip config debug
env:
global:
  /Library/Application Support/pip/pip.conf, exists: False 
site:
  /Users/devesh/pip/.env/pip.conf, exists: True 
    install.timeout: 10 
user:
  /Users/devesh/.pip/pip.conf, exists: False 
  /Users/devesh/.config/pip/pip.conf, exists: True 
    global.timeout: 15 
    global.index-url: https://private_pypi.com/simple/ 
```

Hi @xavfernandez ,

The only thing missing in the output in your example which I implemented in the latest commit is `PIP_INDEX_URL: https://test.pypi.org/simple/` under `env:` in the original example. Where do I look for this key-value pair to be shown under env?

deveshks commented:
Hi @xavfernandez , @pradyunsg 

Any suggestions on the next steps to do here, in order to move this PR ahead and get this approved/merged, especially w.r.t https://github.com/pypa/pip/pull/8096#issuecomment-623009524 are welcome.

pradyunsg commented on news/6741.feature:
```diff
@@ -0,0 +1 @@
+Add a subcommand ``list-files`` to ``pip config`` to list available configuration files and their location
```
This should use the correct command name.

deveshks replied:
Thanks, also do you have any suggestions on how to show the debug options defined as environment variables, as to what should the key look like in that case.?

I feel that we should be using the [PIP_<UPPER_LONG_NAME>](https://pip.pypa.io/en/stable/user_guide/#environment-variables) for it.

deveshks replied:
The command has been renamed in https://github.com/pypa/pip/pull/8096/commits/554a8451703d6d16ae052c20cd85c2dfbf995d49 to `debug` as well.

deveshks replied:
Okay, the output for `env` and `env-var` is now 

```
env_var:
  PIP_CONFIG_FILE='/Users/devesh/Desktop/pip.config'
  PIP_DEFAULT_TIMEOUT='60'
env:
  /Users/devesh/Desktop/pip.config, exists: True 
    global.timeout: 60 
    freeze.timeout: 10 
```

added in https://github.com/pypa/pip/pull/8096/commits/30bc04ac2ad4c5ddfdc4f13c1271b1d69dfdcd94

pradyunsg commented on news/6741.feature:
```diff
@@ -0,0 +1 @@
+Add a subcommand ``debug`` to ``pip config`` to list available configuration files and the key-value pairs defined in them
```
```suggestion
Add a subcommand ``debug`` to ``pip config`` to list available configuration sources and the key-value pairs defined in them
```

deveshks replied:
Fixed in https://github.com/pypa/pip/pull/8096/commits/aced666de85be927e72355a4705829fa21cc1671

deveshks commented:
I have also gone ahead and added tests for this feature for everything except to test for  system-wide configuration file which falls under `global`.

I think we can neither write to system-wide files which might have permissions defined in a way that the tox virtualenv cannot write to those locations, and we cannot patch those files as well since `pip config` commands runs inside a subprocess.

Any suggestions on if there are other ways to achieve this is welcome.

The current output for the command can be found at the PR description.

deveshks commented:
Hi @pradyunsg 

This should be good to go except a test for verifying set global options inside a virtualenv, which I am unsure how to do (https://github.com/pypa/pip/pull/8096#issuecomment-636830994).

Also the Travis CI has passed, but the status is stuck at Pending 😞

deveshks commented:
I added a test in https://github.com/pypa/pip/pull/8096/commits/84d671927ddab3806b725604cc1334d6fd1455af which verifies that the global configuration file is correctly identified for different platforms.

With that, I think that this PR is implementation complete, and should be ready for further reviews.

deveshks commented:
Hi @xavfernandez 

I think this PR is feature complete, and I would really appreciate if you could take a look at this, and provide further reviews.

xavfernandez commented on tests/functional/test_configuration.py:
```diff
@@ -62,3 +67,89 @@ def test_forget_section(self, script):
+
+    def test_user_values(self, script,):
+        """Test that the user pip configuration set using --user
+        is correctly displayed under "user".  This configuration takes place
```
```suggestion
        is correctly displayed under "user". This configuration takes place
```

xavfernandez commented on tests/functional/test_configuration.py:
```diff
@@ -62,3 +67,89 @@ def test_forget_section(self, script):
+        assert re.search(r"site:\n(  .+\n)+", result.stdout)
+
+    def test_global_config_file(self, script):
+        """Test that the system-wide  configuration can be identified"""
```
```suggestion
        """Test that the system-wide configuration can be identified"""
```

xavfernandez commented on tests/functional/test_configuration.py:
```diff
@@ -62,3 +67,89 @@ def test_forget_section(self, script):
+    def test_global_config_file(self, script):
+        """Test that the system-wide  configuration can be identified"""
+
+        # We cannot  write to system-wide files which might have permissions
```
```suggestion
        # We cannot write to system-wide files which might have permissions
```

xavfernandez commented on tests/functional/test_configuration.py:
```diff
@@ -62,3 +67,89 @@ def test_forget_section(self, script):
+        """Test that the system-wide  configuration can be identified"""
+
+        # We cannot  write to system-wide files which might have permissions
+        # defined in a way that the tox virtualenvcannot write to those
```
```suggestion
        # defined in a way that the tox virtualenv cannot write to those
```

deveshks commented:
Thanks @xavfernandez . This should be ready to merge.

> Ideally, we should also help identify invalid/unknown options or environment variable, something like:
> 
> ```
> $ PIP_TOTO=titi pip config debug
> env_var:
>   PIP_TOTO='titi'
> env:
> global:
>   /etc/xdg/pip/pip.conf, exists: False
>   /etc/pip.conf, exists: False
> site:
>   /home/xafer/.virtualenvs/pip/pip.conf, exists: False
> user:
>   /home/xafer/.pip/pip.conf, exists: True
>     global.disable-pip-version-check: 1
>     global.timout: 10
> 
> PIP_TOTO is not a known configuration environment variable
> global.timout is not a known configuration key (did you mean global.timeout ?)
> ```
> 
> but this can definitely happen in a followup PR :)

That's a very good idea, and I think this can definitely be handled in a separate issue/PR.

deveshks commented:
Hi @pradyunsg 

We should be good to merge this unless there are more reviews needed here, post this I will work on https://github.com/pypa/pip/pull/8096#issuecomment-650480588

xavfernandez commented:
Thanks a lot @deveshks :+1:
